File: src/types/vocabulary.ts
/**
 * Shared lookup tables (tags, species, occupations, …). Stored in the
 * project's SQLite database (`vocabularies.db`).
 *
 * Renaming a `VocabularyItem` propagates everywhere automatically
 * because referencing entries store the immutable `id`, not the label.
 */
export type Vocabulary = {
  id: string;
  name: string;
  createdAt: string;
  updatedAt: string;
};

export type VocabularyItem = {
  id: string;
  vocabularyId: string;
  label: string;
  createdAt: string;
  updatedAt: string;
};

/**
 * Id of the project-wide tags vocabulary that every project ships with.
 * Mirrored in src-tauri/src/domain/vocabulary.rs.
 */
export const SYSTEM_TAGS_VOCAB_ID = "tags";

File: src-tauri/src/domain/vocabulary.rs
use serde::{Deserialize, Serialize};

/// System tags vocabulary id — created in every project at scaffolding
/// (or first open). Mirror of `SYSTEM_TAGS_VOCAB_ID` in
/// `src/types/vocabulary.ts`.
pub const SYSTEM_TAGS_VOCAB_ID: &str = "tags";
pub const SYSTEM_TAGS_VOCAB_NAME: &str = "Tags";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Vocabulary {
    pub id: String,
    pub name: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VocabularyItem {
    pub id: String,
    pub vocabulary_id: String,
    pub label: String,
    pub created_at: String,
    pub updated_at: String,
}
